File: hh-examples/11-annotation-usertype-eclipse/src/main/java/com/oreilly/hh/QueryTest2.java
package com.oreilly.hh;

import org.hibernate.query.Query;
import org.hibernate.*;

import com.oreilly.hh.data.*;

import java.time.LocalTime;
import java.util.*;
import java.awt.*;
import java.awt.event.*;
import javax.swing.*;

/**
 * Provide a user interface to enter artist names and see their tracks.
 */
public class QueryTest2 extends JPanel {

    JList<String> list;  // Will contain tracks associated with current artist
    DefaultListModel<String> model; // Lets us manipulate the list contents

    /**
     * Build the panel containing UI elements
     */
    public QueryTest2() {
        setLayout(new BorderLayout());
        model = new DefaultListModel<String>();
        list = new JList<String>(model);
        add(new JScrollPane(list), BorderLayout.SOUTH);

        final JTextField artistField = new JTextField(28);
        artistField.addKeyListener(new KeyAdapter() {
            public void keyTyped(KeyEvent e) {
                SwingUtilities.invokeLater(new Runnable() {
                    public void run() {
                        updateTracks(artistField.getText());
                    }
                });
            }
        });
        add(artistField, BorderLayout.EAST);
        add(new JLabel("Artist: "), BorderLayout.WEST);
    }

    /**
     * Update the list to contain the tracks associated with an artist
     */
    private void updateTracks(String name) {
        model.removeAllElements();  // Clear out previous tracks
        if (name.length() < 1) return;   // Nothing to do

        try {
            Session     session = sessionFactory.openSession();
            Transaction tx = null;
            try {
                tx = session.beginTransaction();

                Artist artist = CreateTest.getArtist(name, false, session);
                if (artist == null) {  // Unknown artist
                    model.addElement("Artist not found");
                } else {                                // yckim
                    // List the tracks associated with the artist
                    for (Track aTrack : artist.getTracks()) {
                        model.addElement(
                            String.format("Track: \"%s\" %s\n", aTrack.getTitle(), aTrack.getPlayTime())
                        );
                    }
                }

                tx.commit();
            } catch (RuntimeException e) {
                if (tx != null) tx.rollback();
                throw e;
            } finally {
                session.close();
            }
        } catch (Exception e) {
            System.err.println("Problem updating tracks:" + e);
            e.printStackTrace();
        }
    }

    private static SessionFactory sessionFactory;

    /**
     * Set up Hibernate, then build and display the user interface.
     */
    public static void main(String args[]) throws Exception {

        sessionFactory = HibernateUtil5.getSessionFactory();

        // Set up the UI
        JFrame frame = new JFrame("Artist Track Lookup");
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.setContentPane(new QueryTest2());
        frame.setSize(400, 180);
        frame.setVisible(true);
    }
}

File: hh-examples/04-3-collection-jpa-kotlin/src/main/kotlin/com/oreilly/hh/QueryTest2.kt
package com.oreilly.hh

import javax.persistence.*

import java.awt.*
import java.awt.event.*
import javax.swing.*

class QueryTest2 : JPanel() {

    var model: DefaultListModel<String> = DefaultListModel()   // Lets us manipulate the list contents
    var list: JList<String> = JList(model)       // Will contain tracks associated with current artist

    /**
     * Build the panel containing UI elements
     */
    init {
        layout = BorderLayout()
        add(JScrollPane(list), BorderLayout.SOUTH)

        val artistField = JTextField(28)
        artistField.addKeyListener(object : KeyAdapter() {
            override fun keyTyped(e: KeyEvent?) {
                SwingUtilities.invokeLater { updateTracks(artistField.text) }
            }
        })
        add(artistField, BorderLayout.EAST)
        add(JLabel("Artist: "), BorderLayout.WEST)
    }

    /**
     * Update the list to contain the tracks associated with an artist
     */
    private fun updateTracks(name: String) {
        model.removeAllElements()       // Clear out previous tracks
        if (name.isEmpty()) return      // Nothing to do

        emf.createEntityManager().use { em ->
            em.transaction.use {
                val artist = getArtist(name, false, em)
                if (artist == null) {  // Unknown artist
                    model.addElement("Artist not found")
                } else {
                    // List the tracks associated with the artist
                    for (aTrack in artist.tracks) {
                        model.addElement("Track: '${aTrack.title}' ${aTrack.playTime}\n")
                    }
                }
            }
        }
    }

    companion object {
        val emf: EntityManagerFactory = Persistence.createEntityManagerFactory("default")
    }
}

fun main() {
    // Set up the UI
    val frame = JFrame("Artist Track Lookup")
    frame.defaultCloseOperation = JFrame.EXIT_ON_CLOSE
    frame.contentPane = QueryTest2()
    frame.setSize(400, 180)
    frame.isVisible = true
}
